packages/@react-spectrum/s2/src/Tabs.tsx

```diff
@@ -10,9 +10,289 @@
+    isHovered: 'neutral-subdued',
+    isDisabled: 'disabled'
+  },
+  borderRadius: 'sm',
+  gap: size(6),
+  height: {
+    density: {
+      compact: 32,
+      regular: 48
+    }
+  },
+  alignItems: 'center',
+  position: 'relative'
+});
+
+const icon = style({
+  paddingEnd: 4,
+  '--iconPrimary': {
+    type: 'fill',
+    value: 'currentColor'
+  }
+});
+
+export function Tab(props: TabProps) {
+  let {density} = useContext(TabsInternalContext);
+
+  return (
+    <RACTab 
+      {...props}
+      className={renderProps => tab({...renderProps, density})}>
+      <Provider
+        values={[
+          [IconContext, {styles: icon}]
+        ]}>
+        {props.children}
+      </Provider>
+    </RACTab>
+  );
+}
+
+const tablist = style({
+  display: 'flex',
+  height: {
+    orientation: {
+      horizontal: {
+        density: {
+          compact: 32,
+          regular: 48
+        }
+      }
+    }
+  },
+  gap: {
+    orientation: {
+      horizontal: {
+        density: {
+          compact: 24,
+          regular: 32
+        }
+      }
+    }
+  },
+  flexDirection: {
+    orientation: {
+      vertical: 'column'
+    }
+  },
+  paddingEnd: {
+    orientation: {
+      vertical: 20
+    }
+  },
+  paddingStart: {
+    orientation: {
+      vertical: 12
+    }
+  },
+  borderColor: 'gray-400'
+});
+
+export function TabList<T extends object>(props: TabListProps<T>) {
+  let {density, isDisabled, disabledKeys, orientation = 'horizontal'} = useContext(TabsInternalContext);
+  let state = useContext(TabListStateContext);
+  let [selectedTab, setSelectedTab] = useState<HTMLElement | undefined>(undefined);
+  let tablistRef = useRef<HTMLDivElement>(null);
+
+  useEffect(() => {    
+    if (tablistRef?.current) {
+      let tab: HTMLElement | null = tablistRef.current.querySelector('[role=tab][data-selected=true]');
+
+      if (tab != null) {
+        setSelectedTab(tab);
+      }
+    }
+
+  }, [tablistRef, state?.selectedItem]);
+
+  return (
+    <div className={style({position: 'relative'})}>
+      {orientation === 'vertical' && 
+        <TabLine disabledKeys={disabledKeys} isDisabled={isDisabled} selectedTab={selectedTab} orientation={orientation} />}
+      <RACTabList 
+        {...props}
+        ref={tablistRef}
+        className={renderProps => tablist({...renderProps, density})} />
+      {orientation === 'horizontal' && 
+        <TabLine disabledKeys={disabledKeys} isDisabled={isDisabled} selectedTab={selectedTab} orientation={orientation} />}
+    </div>
+  );
+}
+
+function isAllTabsDisabled<T>(collection: Collection<T> | null, disabledKeys: Set<Key>) {
+  let selectedKey: Key | null = null;
+  if (collection) {
+    if (disabledKeys.size < collection.size) {
+      return false;
+    }
+    selectedKey = collection.getFirstKey();
+    
+    let index = 0;
+    while (index < collection.size) {
+      if (selectedKey && !disabledKeys.has(selectedKey)) {
+        return false;
+      }
+
+      // @ts-ignore
```

Comment:
why the ts-ignore?

Reply:
originally it was because i thought it would be solved with fixing ts-strict but now that i'm looking at it more i think i need to update the logic